### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by moira2k**
develop tab with adaptive card app

### CODE REVIEW

File: TodoList/bot/api/graphClient.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import "isomorphic-fetch";
+import { TeamsFx, createMicrosoftGraphClient, IdentityType } from "@microsoft/teamsfx";
+import { Client, ResponseType }from "@microsoft/microsoft-graph-client"
+
+interface GraphResponse {
+    status: number;
+    body: { [key: string]: any };
+}
+
+export default async function graphRun(AADId: string, token: string) {
```
Comment (blackchoey) [inline TodoList/bot/api/graphClient.ts]: change funtion name to getUserPhoto?
Also `AADId` parameter can be changed to `aadObjectId`

Comment (blackchoey) [reply TodoList/bot/api/graphClient.ts]: The response type can also be updated

File: TodoList/bot/api/graphClient.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    // Initialize response.
+    const res: GraphResponse = {
+        status: 200,
+        body: {},
+    };
+    if (!token || !token.trim()) {
+        throw new Error('SimpleGraphClient: Invalid token received.');
```
Comment (blackchoey) [inline TodoList/bot/api/graphClient.ts]: Can change error message to `Empty token received` to better describe the problem

File: TodoList/bot/api/graphClient.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    }
+    try {
+        // Get an Authenticated Microsoft Graph client using the token issued to the user.
+        const graphClient = Client.init({
+            authProvider: (done) => {
+                done(null, token); // First parameter takes an error if you can't get an access token.
+            },
+        }); 
+        res.body.profile = await graphClient.api(`/users/${AADId}`).get();
```
Comment (blackchoey) [inline TodoList/bot/api/graphClient.ts:27]: seems we don't need to call this api?

File: TodoList/bot/api/handleData.ts [hunk 4]
```diff
@@ -0,0 +1,245 @@
+import dbRun from "./databaseClient";
+import graphRun from "./graphClient";
+
+interface TodoListResponse {
+    message: string;
+    enquirer: { [key: string]: any };
+    enquiredDate: string;
+    tasks: { [key: string]: any }[];
```
Comment (blackchoey) [inline TodoList/bot/api/handleData.ts]: can only keep tasks

File: TodoList/bot/api/handleData.ts [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        enquiredDate: getCurrentTimeString(),
+        enquirer: {},
+        tasks: []
+    };
+
+    const query:string = `SELECT taskId, taskTime, taskStatus, taskContent, creatorAADId
+    FROM Todo.Tasks WHERE creatorAADId = '${AADId}'`;
+    const req = await dbRun(query);
+    if (req.status != 200) {
+        todoListResp.message = "failure"
```
Comment (blackchoey) [inline TodoList/bot/api/handleData.ts]: improve error handling

File: TodoList/bot/api/handleData.ts [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+
+    var taskMap = {};
+    for (let i:number = 0; i < req.body.content.length; ++i) {
+        if (!(req.body.content[i].taskId in taskMap)) {
+            taskMap[req.body.content[i].taskId] = {
+                taskId: req.body.content[i].taskId,
+                taskTime: req.body.content[i].taskTime,
+                taskStatus: req.body.content[i].taskStatus,
+                taskContent: req.body.content[i].taskContent,
+                creator: creator
```
Comment (blackchoey) [inline TodoList/bot/api/handleData.ts]: may remove this information

File: TodoList/bot/api/handleData.ts [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+
+    return todoListResp;
+}
+
+interface TodoItemResponse {
+    message: string;
+    enquirer: { [key: string]: any };
+    enquiredDate: string;
+    task: { [key: string]: any };
+}
```
Comment (blackchoey) [inline TodoList/bot/api/handleData.ts]: define task
```
interface TodoItemResponse {
    title: String
    time: Date
    sharedWith: string[]
}
```

File: TodoList/bot/api/handleData.ts [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    FROM (SELECT taskId, taskTime, taskStatus, taskContent, creatorAADId FROM Todo.Tasks WHERE taskId = ${taskId}) AS Task
+        LEFT JOIN (SELECT taskId, userAADId FROM Todo.SharedTabs WHERE taskId = ${taskId}) AS A ON Task.taskId = A.taskId`;
+    
+    const req = await dbRun(query);
+    if (req.status != 200) {
+        todoItemResp.message = "failure"
+        return todoItemResp;
+    }
+
+    const creator = await getUserDetails(req.body.content[0].creatorAADId, token);
```
Comment (blackchoey) [inline TodoList/bot/api/handleData.ts]: may remove this

File: TodoList/bot/api/handleData.ts [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+        taskStatus: req.body.content[0].taskStatus,
+        taskContent: req.body.content[0].taskContent,
+        creator: creator,
+        viewers: []
+    }
+
+    var viewerMap = {};
+    for (let i:number = 0; i < req.body.content.length; ++i) {
+        if (req.body.content[i].viewerAADId != null && !(req.body.content[i].viewerAADId in viewerMap)) {
+            let viewer =  await getUserDetails(req.body.content[i].viewerAADId, token);
```
Comment (blackchoey) [inline TodoList/bot/api/handleData.ts]: can do this parallelly

File: TodoList/bot/api/handleData.ts [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+            viewerMap[req.body.content[i].viewerAADId] = 1;
```
Comment (blackchoey) [inline TodoList/bot/api/handleData.ts]: deduplicate during insert

File: TodoList/bot/api/handleData.ts [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+            todoItemResp.task.viewers.push(viewer);
+        }
+    }
+
+    return todoItemResp;
+}
+
+export const handleTodoListAction = async(AADId: string, data: any) => {
```
Comment (blackchoey) [inline TodoList/bot/api/handleData.ts]: async function addTodoItem(objectId: String, task:Task)

File: TodoList/bot/api/handleData.ts [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+    console.log("Handling the Action Of MyTab.");
+    var query:string;
+    switch (data.action) {
+        case "add": {
+            const addTime = data.addDate + "T00:00:00Z";
+            query = `INSERT INTO Todo.Tasks (taskTime, taskStatus, taskContent, creatorAADId) VALUES ('${addTime}', 'Not Started', '${data.addContent}', '${AADId}')`;
```
Comment (blackchoey) [inline TodoList/bot/api/handleData.ts]: taskTime -> dueDate

File: TodoList/bot/api/handleData.ts [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+    console.log("Reading Tasks Shared From Azure SQL Database.");
+    const taskSharedResp:  TodoListResponse = {
+        message: "success",
+        enquiredDate: getCurrentTimeString(),
+        enquirer: {},
+        tasks: []
+    };
+
+    const query:string = `SELECT Todo.Tasks.taskId, Todo.Tasks.taskTime, Todo.Tasks.taskStatus, Todo.Tasks.taskContent, Todo.Tasks.creatorAADId
+    FROM Todo.Tasks INNER JOIN (SELECT taskId FROM Todo.SharedTabs WHERE userAADId = '${AADId}') AS A ON Todo.Tasks.taskId = A.taskId`;
```
Comment (blackchoey) [inline TodoList/bot/api/handleData.ts]: use parameters in query

File: TodoList/bot/teamsBot.ts [hunk 14]
```diff
@@ -1,241 +1,269 @@
...
+    constructor() {
+        super();
+
+        // Creates a new user property accessor.
+        // See https://aka.ms/about-bot-state-accessors to learn more about the bot state and state accessors.
+        // this.userConfigurationProperty = userState.createProperty(
+        //     USER_CONFIGURATION
+        // );
+        // this.userState = userState;
+        this.connectionName = "todolist_v4";
```
Comment (blackchoey) [inline TodoList/bot/teamsBot.ts]: make the value as a config

File: TodoList/bot/teamsBot.ts [hunk 15]
```diff
@@ -1,241 +1,269 @@
...
+            tab: {
+                type: "continue",
+                value: {
+                    cards: []
+                },
+            },
+            responseType: "tab"
+        };
+
+        const user = await getUserDetails(context.activity.from.aadObjectId, tokenResponse.token);
```
Comment (blackchoey) [inline TodoList/bot/teamsBot.ts]: may remove this

File: TodoList/bot/teamsBot.ts [hunk 16]
```diff
@@ -1,241 +1,269 @@
...
+
+        const headingTemplate = new ACData.Template(rawheadingCard);
+        const headingPayload = headingTemplate.expand({
+            $root: {enquirer: user, enquiredDate: getCurrentTimeString()}
+        });
+
+        switch (tabRequest.tabContext.tabEntityId) {
+            // the first tab: MyTab
+            case "MyTab": {
```
Comment (blackchoey) [inline TodoList/bot/teamsBot.ts]: My Todos

File: TodoList/bot/teamsBot.ts [hunk 17]
```diff
@@ -1,241 +1,269 @@
...
+                // Expand the template with your `$root` data object.
+                // This binds it to the data and produces the final Adaptive Card payload
+                const todoListPayload = todoListTemplate.expand({
+                    $root: todoListData
+                });
+                tabFetchResp.tab.value.cards = [{"card": headingPayload}, {"card": todoListPayload}];
+                break;
+            }
+            // the second tab: SharedTab
+            case "SharedTab": {
```
Comment (blackchoey) [inline TodoList/bot/teamsBot.ts]: Shared with me

File: TodoList/createTable.sql [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    taskId INT IDENTITY PRIMARY KEY,
+    taskTime NVARCHAR(128) NOT NULL,
+    taskStatus NVARCHAR(32) NOT NULL default 'Not Started',
+    taskContent NVARCHAR(256) NOT NULL,
+    creatorAADId NVARCHAR(128) NOT NULL,
+);
+
+CREATE TABLE Todo.SharedTabs (
+    Id INT IDENTITY PRIMARY KEY,
+    userAADId NVARCHAR(128) NOT NULL,
```
Comment (blackchoey) [inline TodoList/createTable.sql]: GUID

File: TodoList/createTable.sql [hunk 19]
```diff
@@ -0,0 +1,66 @@
+CREATE SCHEMA Todo;
+
+CREATE TABLE Todo.Tasks (
+    taskId INT IDENTITY PRIMARY KEY,
+    taskTime NVARCHAR(128) NOT NULL,
```
Comment (blackchoey) [inline TodoList/createTable.sql]: datetime

File: TodoList/bot/api/databaseClient.ts [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+
+        return res;
+
+    } catch (err) {
+        console.log("error", err);
+        throw new Error("Failed to execute query in SQL Database.");
+    }
+    // finally {
+    //     if (connection) {
+    //         connection.close();
```
Comment (blackchoey) [inline TodoList/bot/api/databaseClient.ts]: will the connection be reused if we don't close it?

File: TodoList/bot/api/databaseClient.ts [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+    const connection = new Connection(config);
+    connection.connect();
+    return new Promise((resolve, reject) => {
+        connection.on('connect', err => {
+            if (err) {
+                reject(err);
+            }
+            resolve(connection);
+        })
+        // connection.on('debug', function (err) {
```
Comment (blackchoey) [inline TodoList/bot/api/databaseClient.ts]: delete it if we don't need them

File: TodoList/bot/api/databaseClient.ts [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+        //     console.log('debug:', err);
+        // });
+    })
+}
+
+async function execQuery(request: Request, connection: { execSql: (arg0: Request) => void; }) {
```
Comment (blackchoey) [inline TodoList/bot/api/databaseClient.ts]: should `connection: { execSql: (arg0: Request) => void; }` be `connection: Connection` (which is a type from tedious package) ?

File: TodoList/bot/api/graphClient.ts [hunk 23]
```diff
@@ -0,0 +1,39 @@
+// Import polyfills for fetch required by msgraph-sdk-javascript.
+import "isomorphic-fetch";
+import { TeamsFx, createMicrosoftGraphClient, IdentityType } from "@microsoft/teamsfx";
```
Comment (blackchoey) [inline TodoList/bot/api/graphClient.ts]: remove unused imports

File: TodoList/bot/api/graphClient.ts [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+import { Client, ResponseType }from "@microsoft/microsoft-graph-client"
+
+interface UserGraphResponse {
+    profile: { [key: string]: any };
+    profileImage: string;
+}
+
+export default async function graphRun(aadObjectId: string, token: string) {
```
Comment (blackchoey) [inline TodoList/bot/api/graphClient.ts]: may update the function name to better describe what it does

File: TodoList/bot/api/handleData.ts [hunk 25]
```diff
@@ -0,0 +1,378 @@
...
+export async function getUserDetails(aadObjectId: string, token: string) {
+    console.log("Reading User Details From MS Graph.");
+    const user: User = {
+        userName: "",
+        aadObjectId: aadObjectId,
+        profileImage: ""
+    };
+    
+    try {
+        const resp = await graphRun(aadObjectId, token);
```
Comment (blackchoey) [inline TodoList/bot/api/handleData.ts]: may return a `User` instance from graphRun function directly

File: TodoList/bot/api/databaseClient.ts [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+import { TeamsFx, getTediousConnectionConfig } from "@microsoft/teamsfx";
+
+interface DBResponse {
+    body: any;
+}
+
+export default async function dbRun(request: Request): Promise<DBResponse> {
+    // Initialize response.
+    const res: DBResponse = {
+        body: [],
```
Comment (blackchoey) [inline TodoList/bot/api/databaseClient.ts]: maybe no need to wrap the result in DBResponse, can just return an array of rows (the value of body propoerty in current implementation)

File: TodoList/bot/api/handleData.ts [hunk 27]
```diff
@@ -0,0 +1,378 @@
...
+        user.aadObjectId = resp.profile.id;
+        user.profileImage = resp.profileImage;
+
+    } catch (err) {
+        console.log(err);
+    }
+    return user;
+}
+
+interface TodoListResponse {
```
Comment (blackchoey) [inline TodoList/bot/api/handleData.ts]: what's the purpose of define this interface? can we just return `TodoItem[]` in the functions?

File: TodoList/bot/service/botService.ts [hunk 28]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+// Task Module Task Info
+export function createNewItemTaskInfo(content: string = ""): TaskModuleTaskInfo {
+    const taskInfo: TaskModuleTaskInfo = {
+        width: 400,
+    }
+    const time = convertTimeString(new Date());;
+    const newItemTemplate = new ACData.Template(rawNewItemCard);
+    var content:string = repalceHtmlToText(content);
```
Comment (blackchoey) [inline TodoList/bot/service/botService.ts]: why we need to do this?

File: TodoList/bot/teamsBot.ts [hunk 29]
```diff
@@ -1,241 +1,301 @@
...
+import {
+    createAuthResponse,
+    createSignOutResponse,
+    createMyTodosResponse,
+    createSharedwithMeResponse,
+    createNewItemTaskInfo,
+    createActionStatusTaskInfo,
+} from "./service/botService"
+import { User } from "./api/constant";
+// const AdaptiveCardsTools = require("@microsoft/adaptivecards-tools").AdaptiveCards
```
Comment (blackchoey) [inline TodoList/bot/teamsBot.ts]: remove unused comments

File: TodoList/bot/teamsBot.ts [hunk 30]
```diff
@@ -1,241 +1,301 @@
...
+            let txt = context.activity.text;
+            const removedMentionText = TurnContext.removeRecipientMention(context.activity);
+            if (removedMentionText) {
+                // Remove the line break
+                txt = removedMentionText.toLowerCase().replace(/\n|\r/g, "").trim();
+            }
+
+            // Trigger command by IM text
+            switch (txt) {
+                case "welcome": {
```
Comment (blackchoey) [inline TodoList/bot/teamsBot.ts]: do we need this logic?

File: TodoList/bot/teamsBot.ts [hunk 31]
```diff
@@ -1,241 +1,301 @@
...
+            aadObjectId: context.activity.from.aadObjectId,
+            userName: context.activity.from.name
+        };
+
+        let taskInfo: TaskModuleTaskInfo;
+        switch (action.commandId) {
+            case "createTodo": {
+                var content: string;
+                try {
+                    await handleNewItemAction(user.aadObjectId, action.data);
```
Comment (blackchoey) [inline TodoList/bot/teamsBot.ts]: may call the function createToDoItem. By the way, instead ot pass action.data, we can pass actual values (dueDate, content) to the function. So the code may be `await createToDoItem(user.aadObjectId, action.data.dueDate, action.data.content)`

File: TodoList/bot/api/constant.ts [hunk 32]
```diff
@@ -0,0 +1,15 @@
+export interface User {
+    aadObjectId: string;
+    userName: string;
+    profileImage?: string;
+}
+
```
Comment (blackchoey) [inline TodoList/bot/api/constant.ts]: api may not be a good folder name for all the files. may rename the folder or adjust the folder structure for this project

File: TodoList/bot/clients/graphClient.ts [hunk 33]
```diff
@@ -1,7 +1,8 @@
 // Import polyfills for fetch required by msgraph-sdk-javascript.
 import "isomorphic-fetch";
 import { Client, ResponseType }from "@microsoft/microsoft-graph-client"
-import { User } from "./constant";
+import { User } from "../dataModule/user";
```
Comment (blackchoey) [inline TodoList/bot/clients/graphClient.ts]: dataModule -> dataModel

**Review by blackchoey — APPROVED** [review]

**Comment by moira2k** [comment]
basic functions about todoList

**Pull request merged**